iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp

```diff
@@ -45,193 +44,435 @@ namespace iree_compiler {
+    // Convert the function signature to take the HAL ABI LLVM pointers.
     TypeConverter::SignatureConversion signatureConverter(/*numOrigInputs=*/0);
     MLIRContext *context = rewriter.getContext();
-    SmallVector<Type, 5> llvmInputTypes = getABITypes(context);
-    signatureConverter.addInputs(llvmInputTypes);
+    auto abiInputTypes = HALDispatchABI::getInputTypes(context);
+    signatureConverter.addInputs(abiInputTypes);
 
-    // Construct newFunc with all attributes except return type & symbol name.
+    // Copy all attributes onto the LLVM function except the ones handled by
+    // MLIR implicitly.
     SmallVector<NamedAttribute, 4> funcAttrs;
-    for (auto attr : funcOp.getAttrs()) {
+    for (auto attr : stdFuncOp.getAttrs()) {
       if (attr.first == SymbolTable::getSymbolAttrName() ||
           attr.first == mlir::impl::getTypeAttrName()) {
         continue;
       }
       funcAttrs.push_back(attr);
     }
 
+    // Clone the function as an LLVMFuncOp and convert all interior types.
     auto llvmFuncType = LLVM::LLVMFunctionType::get(
-        LLVM::LLVMVoidType::get(rewriter.getContext()), llvmInputTypes);
-    auto newFuncOp = rewriter.create<LLVM::LLVMFuncOp>(
-        funcOp.getLoc(), funcOp.getName(), llvmFuncType,
+        LLVM::LLVMVoidType::get(rewriter.getContext()), abiInputTypes);
+    auto llvmFuncOp = rewriter.create<LLVM::LLVMFuncOp>(
+        stdFuncOp.getLoc(), stdFuncOp.getName(), llvmFuncType,
         LLVM::Linkage::External, funcAttrs);
-
-    // Copy all of funcOp's operations into newFuncOp's body and perform region
-    // type conversion.
-    rewriter.inlineRegionBefore(funcOp.getBody(), newFuncOp.getBody(),
-                                newFuncOp.end());
-    if (failed(rewriter.convertRegionTypes(&newFuncOp.getBody(), *typeConverter,
-                                           &signatureConverter)))
+    rewriter.inlineRegionBefore(stdFuncOp.getBody(), llvmFuncOp.getBody(),
+                                llvmFuncOp.end());
+    if (failed(rewriter.convertRegionTypes(
+            &llvmFuncOp.getBody(), *typeConverter, &signatureConverter))) {
       return failure();
+    }
 
-    rewriter.eraseOp(funcOp);
+    rewriter.eraseOp(stdFuncOp);
     return success();
   }
 };
 
-/// Returns the IREE::HAL::InterfaceBindingOp from an interface op.
-// TODO(ravishankarm) : Copy of similar method in LinalgToSPIRV path. Consider
-// combining them.
-IREE::HAL::InterfaceBindingOp getBindingOp(Operation *op) {
-  if (auto placeholderOp = dyn_cast<IREE::PlaceholderOp>(op)) {
-    return cast<IREE::HAL::InterfaceBindingOp>(
-        SymbolTable::lookupNearestSymbolFrom(
-            op, op->getAttrOfType<SymbolRefAttr>("binding")));
+/// Rewrites hal.interface.workgroup.id to ops loading from the ABI structs.
+///
+/// The parent LLVMFuncOp must be compatible with HALDispatchABI.
+class ConvertHALInterfaceWorkgroupIDOp : public ConvertToLLVMPattern {
```

Comment:
nit: Same as above, these can be template that invokes the correct type(ed) `abi.loadInterface<OpType>`